telemetry/recover_middleware.go

```diff
@@ -0,0 +1,26 @@
+package telemetry
+
+import (
+	"fmt"
+	"net/http"
+	"runtime"
+)
+
+// RecoverMiddleware catches panics and logs them to a span.
+func RecoverMiddleware(next http.Handler) http.Handler {
```

Comment:
we have other error reporting we need to report to also

Reply:
Panics are exported as exceptions on the span (`curl localhost:8111/v1/boom`):

```json
{
	"Name": "/v1/boom",
	"SpanContext": {
		"TraceID": "f19c78cc63bb29dfdb9f27ea15e83635",
		"SpanID": "c9dec87a87940f2e",
		"TraceFlags": "01",
		"TraceState": "",
		"Remote": false
	},
	"Parent": {
		"TraceID": "00000000000000000000000000000000",
		"SpanID": "0000000000000000",
		"TraceFlags": "00",
		"TraceState": "",
		"Remote": false
	},
	"SpanKind": 2,
	"StartTime": "2022-05-03T20:01:44.136725-06:00",
	"EndTime": "2022-05-03T20:01:44.137101416-06:00",
	"Attributes": [
		{
			"Key": "net.transport",
			"Value": {
				"Type": "STRING",
				"Value": "ip_tcp"
			}
		},
		{
			"Key": "net.peer.ip",
			"Value": {
				"Type": "STRING",
				"Value": "::1"
			}
		},
		{
			"Key": "net.peer.port",
			"Value": {
				"Type": "INT64",
				"Value": 64715
			}
		},
		{
			"Key": "net.host.name",
			"Value": {
				"Type": "STRING",
				"Value": "localhost"
			}
		},
		{
			"Key": "net.host.port",
			"Value": {
				"Type": "INT64",
				"Value": 8111
			}
		},
		{
			"Key": "http.method",
			"Value": {
				"Type": "STRING",
				"Value": "GET"
			}
		},
		{
			"Key": "http.target",
			"Value": {
				"Type": "STRING",
				"Value": "/v1/boom"
			}
		},
		{
			"Key": "http.server_name",
			"Value": {
				"Type": "STRING",
				"Value": "the-birds"
			}
		},
		{
			"Key": "http.user_agent",
			"Value": {
				"Type": "STRING",
				"Value": "curl/7.77.0"
			}
		},
		{
			"Key": "http.scheme",
			"Value": {
				"Type": "STRING",
				"Value": "http"
			}
		},
		{
			"Key": "http.host",
			"Value": {
				"Type": "STRING",
				"Value": "localhost:8111"
			}
		},
		{
			"Key": "http.flavor",
			"Value": {
				"Type": "STRING",
				"Value": "1.1"
			}
		},
		{
			"Key": "http.route",
			"Value": {
				"Type": "STRING",
				"Value": "/v1/boom"
			}
		},
		{
			"Key": "http.status_code",
			"Value": {
				"Type": "INT64",
				"Value": 0
			}
		}
	],
	"Events": [
		{
			"Name": "log",
			"Attributes": [
				{
					"Key": "exception.type",
					"Value": {
						"Type": "STRING",
						"Value": "*errors.errorString"
					}
				},
				{
					"Key": "exception.message",
					"Value": {
						"Type": "STRING",
						"Value": "boom\ngoroutine 5 [running]:\ngithub.com/zenledger-io/zazen/telemetry.RecoverMiddleware.func1.1()\n\t/Users/drew/src/zazen/telemetry/recover_middleware.go:19 +0xec\npanic({0x104f54460, 0x1050376d8})\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/runtime/panic.go:838 +0x204\nmain.v1Handler.func2({0x140003c17d0?, 0x104e1292e?}, 0x14000032100?)\n\t/Users/drew/src/zazen/examples/api/main.go:97 +0x30\nnet/http.HandlerFunc.ServeHTTP(0x104f811c0?, {0x10503e418?, 0x14000102120?}, 0x1400003c0d3?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5.(*Mux).routeHTTP(0x140001aaa20, {0x10503e418, 0x14000102120}, 0x14000116900)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 +0x230\nnet/http.HandlerFunc.ServeHTTP(0x140001bd301?, {0x10503e418?, 0x14000102120?}, 0x104b85ac0?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5.(*Mux).ServeHTTP(0x140001aaa20, {0x10503e418, 0x14000102120}, 0x14000116900)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:71 +0x458\ngithub.com/go-chi/chi/v5.(*Mux).Mount.func1({0x10503e418, 0x14000102120}, 0x14000116900)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:314 +0x1c0\nnet/http.HandlerFunc.ServeHTTP(0x104f811c0?, {0x10503e418?, 0x14000102120?}, 0x14000036094?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5.(*Mux).routeHTTP(0x140001aab40, {0x10503e418, 0x14000102120}, 0x14000116900)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 +0x230\nnet/http.HandlerFunc.ServeHTTP(0x0?, {0x10503e418?, 0x14000102120?}, 0x90?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/zenledger-io/zazen/telemetry.RecoverMiddleware.func1({0x10503e418?, 0x14000102120?}, 0x1?)\n\t/Users/drew/src/zazen/telemetry/recover_middleware.go:24 +0x70\nnet/http.HandlerFunc.ServeHTTP(0x10503dae8?, {0x10503e418?, 0x14000102120?}, 0x0?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/riandyrn/otelchi.traceware.ServeHTTP({{0x104e1728e, 0x9}, {0x105038ae8, 0x140001d59c0}, {0x10503d6c8, 0x140001d4f00}, {0x10503b1e8, 0x140001a4ab0}, {0x0, 0x0}, ...}, ...)\n\t/Users/drew/go/pkg/mod/github.com/riandyrn/otelchi@v0.4.0/middleware.go:142 +0x624\ngithub.com/go-chi/chi/v5/middleware.RequestID.func1({0x10503dae8, 0x1400011c000}, 0x14000116200)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/request_id.go:76 +0x344\nnet/http.HandlerFunc.ServeHTTP(0x14000116200?, {0x10503dae8?, 0x1400011c000?}, 0x14000118240?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5/middleware.RealIP.func1({0x10503dae8, 0x1400011c000}, 0x14000116200)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/realip.go:35 +0x98\nnet/http.HandlerFunc.ServeHTTP(0x10503e760?, {0x10503dae8?, 0x1400011c000?}, 0x1057ee3a0?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5/middleware.Timeout.func1.1({0x10503dae8, 0x1400011c000}, 0x14000116100)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/timeout.go:45 +0x2b0\nnet/http.HandlerFunc.ServeHTTP(0x10503e6b8?, {0x10503dae8?, 0x1400011c000?}, 0x1053f3620?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5.(*Mux).ServeHTTP(0x140001aab40, {0x10503dae8, 0x1400011c000}, 0x14000116000)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:88 +0x414\nnet/http.serverHandler.ServeHTTP({0x1400010e0c0?}, {0x10503dae8, 0x1400011c000}, 0x14000116000)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2916 +0x3fc\nnet/http.(*conn).serve(0x14000110000, {0x10503e760, 0x140003c1b60})\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:1966 +0x56c\ncreated by net/http.(*Server).Serve\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:3071 +0x450\n"
					}
				},
				{
					"Key": "log.severity",
					"Value": {
						"Type": "STRING",
						"Value": "ERROR"
					}
				},
				{
					"Key": "log.message",
					"Value": {
						"Type": "STRING",
						"Value": "panic"
					}
				},
				{
					"Key": "code.function",
					"Value": {
						"Type": "STRING",
						"Value": "github.com/zenledger-io/zazen/telemetry.(*otelZapLog).Error"
					}
				},
				{
					"Key": "code.filepath",
					"Value": {
						"Type": "STRING",
						"Value": "/Users/drew/src/zazen/telemetry/zap_log.go"
					}
				},
				{
					"Key": "code.lineno",
					"Value": {
						"Type": "INT64",
						"Value": 34
					}
				}
			],
			"DroppedAttributeCount": 0,
			"Time": "2022-05-03T20:01:44.136978-06:00"
		},
		{
			"Name": "exception",
			"Attributes": [
				{
					"Key": "exception.type",
					"Value": {
						"Type": "STRING",
						"Value": "*errors.errorString"
					}
				},
				{
					"Key": "exception.message",
					"Value": {
						"Type": "STRING",
						"Value": "boom\ngoroutine 5 [running]:\ngithub.com/zenledger-io/zazen/telemetry.RecoverMiddleware.func1.1()\n\t/Users/drew/src/zazen/telemetry/recover_middleware.go:19 +0xec\npanic({0x104f54460, 0x1050376d8})\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/runtime/panic.go:838 +0x204\nmain.v1Handler.func2({0x140003c17d0?, 0x104e1292e?}, 0x14000032100?)\n\t/Users/drew/src/zazen/examples/api/main.go:97 +0x30\nnet/http.HandlerFunc.ServeHTTP(0x104f811c0?, {0x10503e418?, 0x14000102120?}, 0x1400003c0d3?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5.(*Mux).routeHTTP(0x140001aaa20, {0x10503e418, 0x14000102120}, 0x14000116900)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 +0x230\nnet/http.HandlerFunc.ServeHTTP(0x140001bd301?, {0x10503e418?, 0x14000102120?}, 0x104b85ac0?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5.(*Mux).ServeHTTP(0x140001aaa20, {0x10503e418, 0x14000102120}, 0x14000116900)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:71 +0x458\ngithub.com/go-chi/chi/v5.(*Mux).Mount.func1({0x10503e418, 0x14000102120}, 0x14000116900)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:314 +0x1c0\nnet/http.HandlerFunc.ServeHTTP(0x104f811c0?, {0x10503e418?, 0x14000102120?}, 0x14000036094?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5.(*Mux).routeHTTP(0x140001aab40, {0x10503e418, 0x14000102120}, 0x14000116900)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 +0x230\nnet/http.HandlerFunc.ServeHTTP(0x0?, {0x10503e418?, 0x14000102120?}, 0x90?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/zenledger-io/zazen/telemetry.RecoverMiddleware.func1({0x10503e418?, 0x14000102120?}, 0x1?)\n\t/Users/drew/src/zazen/telemetry/recover_middleware.go:24 +0x70\nnet/http.HandlerFunc.ServeHTTP(0x10503dae8?, {0x10503e418?, 0x14000102120?}, 0x0?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/riandyrn/otelchi.traceware.ServeHTTP({{0x104e1728e, 0x9}, {0x105038ae8, 0x140001d59c0}, {0x10503d6c8, 0x140001d4f00}, {0x10503b1e8, 0x140001a4ab0}, {0x0, 0x0}, ...}, ...)\n\t/Users/drew/go/pkg/mod/github.com/riandyrn/otelchi@v0.4.0/middleware.go:142 +0x624\ngithub.com/go-chi/chi/v5/middleware.RequestID.func1({0x10503dae8, 0x1400011c000}, 0x14000116200)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/request_id.go:76 +0x344\nnet/http.HandlerFunc.ServeHTTP(0x14000116200?, {0x10503dae8?, 0x1400011c000?}, 0x14000118240?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5/middleware.RealIP.func1({0x10503dae8, 0x1400011c000}, 0x14000116200)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/realip.go:35 +0x98\nnet/http.HandlerFunc.ServeHTTP(0x10503e760?, {0x10503dae8?, 0x1400011c000?}, 0x1057ee3a0?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5/middleware.Timeout.func1.1({0x10503dae8, 0x1400011c000}, 0x14000116100)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/middleware/timeout.go:45 +0x2b0\nnet/http.HandlerFunc.ServeHTTP(0x10503e6b8?, {0x10503dae8?, 0x1400011c000?}, 0x1053f3620?)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2084 +0x3c\ngithub.com/go-chi/chi/v5.(*Mux).ServeHTTP(0x140001aab40, {0x10503dae8, 0x1400011c000}, 0x14000116000)\n\t/Users/drew/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:88 +0x414\nnet/http.serverHandler.ServeHTTP({0x1400010e0c0?}, {0x10503dae8, 0x1400011c000}, 0x14000116000)\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:2916 +0x3fc\nnet/http.(*conn).serve(0x14000110000, {0x10503e760, 0x140003c1b60})\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:1966 +0x56c\ncreated by net/http.(*Server).Serve\n\t/Users/drew/.homebrew/Cellar/go/1.18/libexec/src/net/http/server.go:3071 +0x450\n"
					}
				}
			],
			"DroppedAttributeCount": 0,
			"Time": "2022-05-03T20:01:44.137098-06:00"
		}
	],
	"Links": null,
	"Status": {
		"Code": "Error",
		"Description": "Invalid HTTP status code 0"
	},
	"DroppedAttributes": 0,
	"DroppedEvents": 0,
	"DroppedLinks": 0,
	"ChildSpanCount": 0,
	"Resource": [
		{
			"Key": "service.name",
			"Value": {
				"Type": "STRING",
				"Value": "the-birds"
			}
		},
		{
			"Key": "service.version",
			"Value": {
				"Type": "STRING",
				"Value": "0.0.1"
			}
		}
	],
	"InstrumentationLibrary": {
		"Name": "github.com/riandyrn/otelchi",
		"Version": "semver:1.0.0",
		"SchemaURL": ""
	}
}
```

Whatever collector we use can/should forward exceptions to the correct place.